tests/integration_tests/test_wandb.py

```diff
@@ -28,114 +30,267 @@ def _multiobjective_func(trial: optuna.trial.Trial) -> Tuple[float, float]:
     return first_objective, second_objective
 
 
+def logging_objective_func(trial: optuna.trial.Trial, log_func: Callable) -> float:
+    result = _objective_func(trial)
+    log_func({"result": result})
+    return result
+
+
+def logging_multiobjective_func(
+    trial: optuna.trial.Trial, log_func: Callable
+) -> Tuple[float, float]:
+    result0, result1 = _multiobjective_func(trial)
+    log_func({"result0": result0, "result1": result1})
+    return result0, result1
+
+
 @mock.patch("optuna.integration.wandb.wandb")
 def test_run_initialized(wandb: mock.MagicMock) -> None:
 
+    wandb.sdk.wandb_run.Run = mock.MagicMock
+
+    n_trials = 10
     wandb_kwargs = {
         "project": "optuna",
         "group": "summary",
         "job_type": "logging",
         "mode": "offline",
+        "tags": ["test-tag"],
     }
 
-    WeightsAndBiasesCallback(
-        metric_name="mse",
-        wandb_kwargs=wandb_kwargs,
+    WeightsAndBiasesCallback(metric_name="mse", wandb_kwargs=wandb_kwargs, as_multirun=False)
+    wandb.init.assert_called_once_with(
+        project="optuna", group="summary", job_type="logging", mode="offline", tags=["test-tag"]
+    )
+
+    wandbc = WeightsAndBiasesCallback(
+        metric_name="mse", wandb_kwargs=wandb_kwargs, as_multirun=True
     )
+    wandb.run = None
+
+    study = optuna.create_study(direction="minimize")
+    _wrapped_func = wandbc.track_in_wandb(lambda t: 1.0)
+    wandb.init.reset_mock()
+    trial = optuna.create_trial(value=1.0)
+    _wrapped_func(trial)
 
     wandb.init.assert_called_once_with(
-        project="optuna",
-        group="summary",
-        job_type="logging",
-        mode="offline",
+        project="optuna", group="summary", job_type="logging", mode="offline", tags=["test-tag"]
+    )
+
+    wandb.init.reset_mock()
+    study.optimize(_objective_func, n_trials=n_trials, callbacks=[wandbc])
+
+    wandb.init.assert_called_with(
+        project="optuna", group="summary", job_type="logging", mode="offline", tags=["test-tag"]
     )
 
+    assert wandb.init.call_count == n_trials
+
+    wandb.init().finish.assert_called()
+    assert wandb.init().finish.call_count == n_trials
+
 
 @mock.patch("optuna.integration.wandb.wandb")
-def test_attributes_set_on_epoch(wandb: mock.MagicMock) -> None:
+@pytest.mark.parametrize("as_multirun", [True, False])
+def test_attributes_set_on_epoch(wandb: mock.MagicMock, as_multirun: bool) -> None:
 
-    wandb.config.update = mock.MagicMock()
+    # Vanilla update
+    wandb.sdk.wandb_run.Run = mock.MagicMock
+    expected = {"direction": ["MINIMIZE"]}
+    trial_params = {"x": 1.1, "y": 2.2}
+    expected_with_params = {"direction": ["MINIMIZE"], "x": 1.1, "y": 2.2}
```

Comment:
```suggestion
    expected_config_with_params = expected_config | trial_params
```

Reply:
oh, nice comment. I didn't know that.

Reply:
Don’t think we can. MyPy complains if I do something like:
```
expected.update(trial_params)
```
or
```
expected_config_with_params = {**expected, **trial_params}
```
That’s because, all the entries of `expected` are strings, but the values of `trial_params` are floating point numbers, and mypy complains if we dynamically try to create a heterogeneous dictionary.

I thought maybe we could use `TypedDict` for this, but importing `TypedDict` is different for python 3.9 and other lower versions. Do you guys have any suggestions?

Reply:
How about declaring a type for `expected`? The test becomes:
```python
from typing import Any


@mock.patch("optuna.integration.wandb.wandb")
@pytest.mark.parametrize("as_multirun", [True, False])
def test_attributes_set_on_epoch(wandb: mock.MagicMock, as_multirun: bool) -> None:

    # Vanilla update.
    wandb.sdk.wandb_run.Run = mock.MagicMock
    expected: Dict[str, Any] = {"direction": ["MINIMIZE"]}
    trial_params = {"x": 1.1, "y": 2.2}

    study = optuna.create_study(direction="minimize")
    wandbc = WeightsAndBiasesCallback(as_multirun=as_multirun)
    study.enqueue_trial(trial_params)
    study.optimize(_objective_func, n_trials=1, callbacks=[wandbc])

    if as_multirun:
        wandb.run = None
        expected.update(trial_params)
        study.enqueue_trial(trial_params)
        study.optimize(_objective_func, n_trials=1, callbacks=[wandbc])
        wandb.init().config.update.assert_called_once_with(expected)
    else:
        wandb.run.config.update.assert_called_once_with(expected)
```